### PR TITLE
Add Remove and RemoveNull filter rules

### PR DESCRIPTION
**Pull request by rick-nu**
### What

Removing values from the input array requires some core refactoring.
### How to test
1. See that there are tests that can remove values form the to filter array
2. See that the test coverage is still 100% and the quality a 10
### ToDo
- [x] Find a way to properly unset values from the data container
- [x] Add decent tests for the removeNull filter rule
- [x] Fix removal on multidimensional arrays
- [x] Documentation
- [x] Extend readme example
- [x] ~~Code quality 10~~ Make it as clean as possible
### Ticket

 #42

### CODE REVIEW

**Comment by jhuet** [comment]
Wow, thanks ! I didn't expect my demand to be such a serious thing :) Can't wait to try it !

**Comment by rick-nu** [comment]
@jhuet Haha no problem, like I said, it's a great feature request :D I think I have a pretty decent solution now. What do you think, can find anything to improve?

File: src/Chain.php [hunk 1]
```diff
@@ -23,23 +25,33 @@ class Chain
...
         foreach ($this->rules as $rule) {
             $rule->setFilterData($filterData);
-            if ($isSet || $rule->allowedNotSet()) {
-                $value = $rule->filter($value);
-                $isSet = true;
+            if ($isNotEmpty || $rule->allowedNotSet()) {
+                $filterResult = $rule->filter($value);
+
+                if (!$filterResult instanceof FilterResult) {
+                    throw new Exception(
```
Comment (berry-langerak) [inline src/Chain.php]: Why not a specific exception instead? I don't like using generic Exceptions in a third-party library (harder to catch, and all ;))

Comment (rick-nu) [inline src/Chain.php]: Then [LogicException](http://www.php.net/manual/en/class.logicexception.php) sounds like the best option to me, agreed?

File: src/Value/Container.php [hunk 2]
```diff
@@ -118,4 +128,26 @@ protected function setTraverse($key, $value)
...
+     * @param array $values
+     */
+    private function removeRecursive(array $keyParts, array &$values)
+    {
+        $key = $keyParts[0];
+        if (array_key_exists($key, $values)) {
+            if (is_array($values[$key])) {
+                $this->removeRecursive(array_splice($keyParts, 1), $values[$key]);
+
+                // unset self if the removed child clears this array
```
Comment (berry-langerak) [inline src/Value/Container.php]: if if if you need inline comments, you probably need to abstract into another method.

File: src/Chain.php (hunk 1 above)
Comment (rick-nu) [inline src/Chain.php]: Or actually extend the Exception class is what you mean

File: src/Value/Container.php (hunk 2 above)
Comment (berry-langerak) [inline src/Value/Container.php]: oh, and the if if if was a joke, but the nesting of this method (purely based on trying to get it) is a bit much, I think.

File: src/Chain.php (hunk 1 above)
Comment (berry-langerak) [inline src/Chain.php]: Actually extend the `Exception` class.

File: src/Value/Container.php (hunk 2 above)
Comment (rick-nu) [inline src/Value/Container.php]: Tried that, but for some reasons that breaks the reference to the object, and the unset will fail

**Comment by berry-langerak** [comment]
@RickvdStaaij looks pretty good, just a few comments on complexity of methods (try to keep it readable, for your future self ;)). Also, I wonder if returning a `FilterResult` from the concrete filters is actually necessary. Can't that be abstracted somehow?

**Comment by rick-nu** [comment]
@berry-langerak I found a way that FilterResult is not required to be resulted in a FilterRule! :D A billion times better if you ask me.

**Comment by rick-nu** [comment]
100% coverage, code quality 10, not going for any less :wink:

**Pull request merged**